src/osm4routing/reader.rs

```diff
@@ -1,259 +1,434 @@
+
+    // remove ways with missing nodes
+    for way_index in ways_to_remove.iter().rev() {
+        self.ways.remove(*way_index);
     }
 
-    fn read_ways(&mut self, file: std::fs::File) {
-        let mut pbf = osmpbfreader::OsmPbfReader::new(file);
-        for obj in pbf.iter().flatten() {
-            if let osmpbfreader::OsmObj::Way(way) = obj {
-                let mut skip = false;
-                let mut properties = EdgeProperties::default();
-                for (key, val) in way.tags.iter() {
-                    properties.update(key.to_string(), val.to_string());
-                    if self
-                        .forbidden
-                        .get(key.as_str())
-                        .map(|vals| vals.contains(val.as_str()) || vals.contains("*"))
-                        == Some(true)
-                    {
-                        skip = true;
-                    }
-                }
-                properties.normalize();
-                if properties.accessible() && !skip {
-                    for node in &way.nodes {
-                        self.nodes_to_keep.insert(*node);
-                    }
-                    self.ways.push(Way {
-                        id: way.id,
-                        nodes: way.nodes,
-                        properties,
-                    });
-                }
-            }
-        }
+    // print stats
+    for (way_index, node_id) in missing_node_stats {
+      println!("Way {} is missing node {}", way_index, node_id);
     }
+    println!("{} ways removed ({:.2}%) because of missing nodes", removed_ways_count, removed_ways_percentage);
+    println!("Missing nodes stats:");
+  }
 
-    fn read_nodes(&mut self, file: std::fs::File) {
-        let mut pbf = osmpbfreader::OsmPbfReader::new(file);
-        self.nodes.reserve(self.nodes_to_keep.len());
-        for obj in pbf.iter().flatten() {
-            if let osmpbfreader::OsmObj::Node(node) = obj {
-                if self.nodes_to_keep.contains(&node.id) {
-                    self.nodes_to_keep.remove(&node.id);
-                    self.nodes.insert(
-                        node.id,
-                        Node {
-                            id: node.id,
-                            coord: Coord {
-                                lon: node.lon(),
-                                lat: node.lat(),
-                            },
-                            uses: 0,
-                        },
-                    );
-                }
-            }
-        }
+  /// Splits a way into multiple edges based on nodes with more than one usage.
+  ///
+  /// # Arguments
+  ///
+  /// * `way` - The way to split into edges.
+  ///
+  /// # Returns
+  ///
+  /// A vector of edges resulting from splitting the way. Each edge represents a segment of the original way
+  /// between nodes with more than one usage. The vector is empty if the way does not need to be split.
+  fn split_way(&self, way: &Way) -> Vec<Edge> {
```

Comment:
Do we use this function right now?

Reply:
yes
https://github.com/Esmartdriver/osm4routing2/blob/803e2d2c0ddb9c896a573a39c0a3a0671b71f635/src/osm4routing/reader.rs#L131